contribs/drt-extensions/src/main/java/org/matsim/contrib/drt/extension/prebooking/dvrp/PrebookingManager.java

```diff
@@ -0,0 +1,164 @@
+import org.matsim.contrib.drt.passenger.AcceptedDrtRequest;
+import org.matsim.contrib.dvrp.optimizer.Request;
+import org.matsim.contrib.dvrp.optimizer.VrpOptimizer;
+import org.matsim.contrib.dvrp.passenger.PassengerRequest;
+import org.matsim.contrib.dvrp.passenger.PassengerRequestCreator;
+import org.matsim.contrib.dvrp.passenger.PassengerRequestRejectedEvent;
+import org.matsim.contrib.dvrp.passenger.PassengerRequestValidator;
+import org.matsim.core.api.experimental.events.EventsManager;
+import org.matsim.core.mobsim.framework.MobsimAgent;
+import org.matsim.core.mobsim.qsim.InternalInterface;
+import org.matsim.core.mobsim.qsim.interfaces.MobsimEngine;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Verify;
+
+/**
+ * This class manages prebooked requests. One instance of PrebookingManager
+ * exists per mode. The entry point is PrebookingManager::prebook to which you
+ * need to pass a person, a leg with the respective DRT mode, the
+ * requested/expected earliest departure time, and the time at which the request
+ * should be submitted / taken into account in the system.
+ * 
+ * Preplanned requests can be submitted any time before the planned
+ * departure/submission times.
+ * 
+ * Internally, the prebooking manager will create a request identifier and
+ * return the request once the agent actually wants to depart on the planned
+ * leg. The link between a leg and a request is managed by inserting a special
+ * attribute in the leg instance.
+ * 
+ * @author Sebastian Hörl (sebhoerl), IRT SystemX
+ */
+public class PrebookingManager implements MobsimEngine {
+	private static final String PREBOOKED_REQUEST_PREFIX = "prebookedRequestId";
+
+	private final String mode;
+
+	private final Network network;
+	private final EventsManager eventsManager;
+
+	private final PassengerRequestCreator requestCreator;
+	private final PassengerRequestValidator requestValidator;
+	private final VrpOptimizer optimizer;
+
+	private final AtomicInteger currentRequestIndex = new AtomicInteger(-1);
+
+	private final String requestAttribute;
+
+	private final List<PrebookingQueueItem> prebookingQueue = new LinkedList<>();
+	private final IdMap<Request, PassengerRequest> prebookedRequests = new IdMap<>(Request.class);
+
+	public PrebookingManager(String mode, Network network, PassengerRequestCreator requestCreator,
+			VrpOptimizer optimizer, PassengerRequestValidator requestValidator, EventsManager eventsManager) {
+		this.network = network;
+		this.mode = mode;
+		this.requestCreator = requestCreator;
+		this.optimizer = optimizer;
+		this.requestAttribute = PREBOOKED_REQUEST_PREFIX + ":" + mode;
+		this.requestValidator = requestValidator;
+		this.eventsManager = eventsManager;
+	}
+
+	PassengerRequest consumePrebookedRequest(MobsimAgent agent, Leg leg) {
+		Verify.verify(leg.getMode().equals(mode), "Invalid mode for this prebooking manager");
```

Comment:
I'm not really sure where to put what, code-wise it doesn't seem to make a difference. Personally I like to use Verify, because then I can just put a breakpoint into Verify.verify and find any eventual errors in scheduling, etc. Do you have any good guidelines? If not I would rather skip this point to quickly bring the PR into a mergeable state :)

Reply:
Here's is a good guideline: https://github.com/google/guava/wiki/ConditionalFailuresExplained. But sometimes it's unclear which way to go.

> If not I would rather skip this point to quickly bring the PR into a mergeable state :)

It's not a blocker, you can ignore my comment :-)